kernel/src/util/ioctl/mod.rs

```diff
@@ -0,0 +1,441 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Decoding and dispatching ioctl commands.
+//!
+//! When the system call layer handles an ioctl system call, it creates a [`RawIoctl`], which is
+//! basically the numeric ioctl command and argument.
+//!
+//! The component that handles the specific ioctl logic should first convert the [`RawIoctl`] to a
+//! strongly typed [`Ioctl`] instance, which decodes the ioctl command and includes type
+//! information about the ioctl argument.
+//!
+//! To achieve this, define type aliases for [`Ioctl`] using the [`ioc`] macro, preferably in a
+//! separate module for clarity. At the function that dispatches the ioctl commands, we can
+//! glob-import the defined [`Ioctl`] type aliases and use the [`ioctl`] macro for the ioctl
+//! dispatching.
+//!
+//! Here is an complete example that demonstrates the basic usage:
+//! ```
+//! mod ioctl_defs {
+//!     use crate::util::ioctl::{ioc, InData, OutData, PassByVal};
+//!
+//!     // Below we give the code in Linux to provide an intuitive guide on how to use the `ioc`
+//!     // macro and how it corresponds to the Linux definitions.
+//!
+//!     // ```c
+//!     // #define TIOCSCTTY    0x540E
+//!     // ```
+//!     pub(super) type SetControlTty = ioc!(TIOCSCTTY, 0x540E, InData<i32, PassByVal>);
+//!
+//!     // ```c
+//!     // #define TIOCSPTLCK   _IOW('T', 0x31, int)  /* Lock/unlock Pty */
+//!     // ```
+//!     pub(super) type SetPtyLock = ioc!(TIOCSPTLCK, b'T', 0x31, InData<i32>);
+//!
+//!     // ```c
+//!     // #define TIOCGPTLCK   _IOR('T', 0x39, int) /* Get Pty lock state */
+//!     //  ```
+//!     pub(super) type GetPtyLock = ioc!(TIOCGPTLCK, b'T', 0x39, OutData<i32>);
+//! }
+//!
+//! #[derive(Debug, Default)]
+//! struct TtyState {
+//!     is_controlling: bool,
+//!     is_locked: bool,
+//! }
+//!
+//! impl TtyState {
+//!     fn ioctl(&mut self, raw_ioctl: RawIoctl) -> Result<()> {
+//!         use ioctl_defs::*;
+//!
+//!         dispatch_ioctl!(match raw_ioctl {
+//!             cmd @ SetControlTty => {
+//!                 let _is_forced = cmd.get() != 0;
+//!                 self.is_controlling = true;
+//!             }
+//!             cmd @ SetPtyLock => {
+//!                 self.is_locked = cmd.read()? != 0;
+//!             }
+//!             cmd @ GetPtyLock => {
+//!                 cmd.write(if self.is_locked { &1 } else { &0 })?;
+//!             }
+//!             _ => return_errno_with_message!(Errno::ENOTTY, "the ioctl command is unknown"),
+//!         });
+//!         Ok(())
+//!     }
+//! }
+//!
+//! fn ioctl_dispatch() {
```

Comment:
I suggest renaming to a more specific and accurate name, `do_tty_ioctl` or `fake_tty_ioctl`, for improved readability.